sgl-kernel/src/sgl-kernel/csrc/cpu/rope.cpp

```diff
@@ -0,0 +1,122 @@
+    auto out_stride_qs = query.stride(0);
+    auto out_stride_ks = key.stride(0);
+    auto emb_pos_sin_ptr = t_emb_pos_sin.data_ptr<T>(); // [MP][1][HR]
+    auto emb_pos_cos_ptr = t_emb_pos_cos.data_ptr<T>(); // [MP][1][HR]
+    {
+#pragma omp parallel for collapse(2)
+    for (int s = 0; s < S; s++) {
+        for (int n = 0; n < N; n++) {
+            auto in_offset_q = s * in_stride_s + n * rotary_dim;
+            auto out_offset_q = s * out_stride_qs + n * rotary_dim;
+            auto out_offset_k = s * out_stride_ks;
+            long p = 0;
+            T* sin_start = nullptr;
+            T* cos_start = nullptr;
+            // step 0) get the rotary position embedding for the current position
+            sin_start = emb_pos_sin_ptr + s * HR;
+            cos_start = emb_pos_cos_ptr + s * HR;
+            // step 1) apply_rotary_pos_emb for the rotary_dim elements in every
+            // head of query/key
+            for (auto h = 0; h < rotary_dim; h += 2) {
+                auto cos1 = cos_start[h];
+                auto sin1 = sin_start[h];
+                auto cos2 = cos_start[h + 1];
+                auto sin2 = sin_start[h + 1];
+                auto in1 = in_ptr[in_offset_q + h];
+                auto in2 = in_ptr[in_offset_q + h + 1];
+                auto out1 = in1 * cos1 - in2 * sin1;
+                auto out2 = in2 * cos2 + in1 * sin2;
+                auto out1_offset = out_offset_q + h;
+                auto out2_offset = out1_offset + 1;
+                query_ptr[out1_offset] = out1;
+                query_ptr[out2_offset] = out2;
+            }
+            if (n < NK) {
+                for (auto h = 0; h < HK; h += 2) {
+                    auto cos1 = cos_start[h];
+                    auto sin1 = sin_start[h];
+                    auto cos2 = cos_start[h + 1];
+                    auto sin2 = sin_start[h + 1];
+                    auto k_pe_offset = s * k_pe_stride_s;
+                    auto in1_k = k_pe_ptr[k_pe_offset + h];
+                    auto in2_k = k_pe_ptr[k_pe_offset + h + 1];
+                    auto out1_k = in1_k * cos1 - in2_k * sin1;
+                    auto out2_k = in2_k * cos2 + in1_k * sin2;
+                    auto out1_offset = out_offset_k + h;
+                    auto out2_offset = out1_offset + 1;
+                    key_ptr[out1_offset] = out1_k;
+                    key_ptr[out2_offset] = out2_k;
+                }
+            }
+        }
+    }
+    }
+    return std::make_tuple(query, key);
+}
+
+}
+
+std::tuple<at::Tensor, at::Tensor>
+rotary_position_embedding_cpu(
+    at::Tensor& q_pe,
+    at::Tensor& k_pe,
+    at::Tensor& t_emb_pos_sin,
+    at::Tensor& t_emb_pos_cos) {
+  RECORD_FUNCTION(
+    "sgl-kernel::rotary_position_embedding_cpu", std::vector<c10::IValue>({q_pe, k_pe, t_emb_pos_sin, t_emb_pos_cos}));
+  q_pe = q_pe.contiguous();
+  k_pe = k_pe.contiguous();
+  t_emb_pos_sin = t_emb_pos_sin.contiguous();
+  t_emb_pos_cos = t_emb_pos_cos.contiguous();
```

Comment:
do we have non-contiguous inputs here?

Reply:
Sin and cos in V2 and V2-lite are all contiguous, q_pe in prefill is **non**-contiguous, k_pe in prefill is contiguous. q_pe and k_pe in decode is all contiguous.